File: 210_Course_Scedule_II/findOrder.py
class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """
        res = []
        graph = {i:set() for i in xrange(numCourses)}
        indeg = {i:0     for i in xrange(numCourses)}

        for s, e in set(tuple(x) for x in prerequisites):
            graph[e] |= {s}
            indeg[s] += 1


        queue = [i for i in xrange(numCourses) if not indeg[i]]
        for node in queue:
            res.append(node)
            for neigh in graph[node]:
                indeg[neigh] -= 1
                if not indeg[neigh]:
                    queue += neigh,

        return res if len(res) == numCourses else []

sol = Solution()
numCourses = 4
prerequisites = [[1,0],[2,0],[3,1],[3,2]]
print sol.findOrder(numCourses, prerequisites)

File: 210_Course_Scedule_II/findOrder.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <utility>

using namespace std;

class Solution {
    public:
        vector<int> findOrder(int numCourses, vector<pair<int, int> >& prerequisites) {
            vector<unordered_set<int> > graph = makeGraph(numCourses, prerequisites);
            vector<int> degree = countInDegree(graph);
            vector<int> res;

            for(int i = 0; i < numCourses; i++) {
                int j = 0;
                for(; j < numCourses; j++) {
                    if(degree[j] == 0) { break; }
                }
                if(j == numCourses) { return vector<int>(); } 
                res.push_back(j);
                degree[j] = -1;
                for(int neigh : graph[j]) {
                    degree[neigh]--;
                }
            }

            return res;

        }

    private:
        vector<unordered_set<int> > makeGraph(int numCourses, vector<pair<int, int> >& prerequisites) {
            vector<unordered_set<int> > graph(numCourses);

            for(auto pre : prerequisites) {
                graph[pre.second].insert(pre.first);
            }

            return graph;
        }

        vector<int> countInDegree(vector<unordered_set<int> >& graph) {
            vector<int> degree(graph.size(), 0);

            for(auto neighbors : graph) {
                for(int neigh : neighbors) {
                    degree[neigh]++;
                }
            }

            return degree;
        }
};
